R/data-combined.R

```diff
@@ -0,0 +1,625 @@
+      # if alternate names are provided, change the names
+      # for `NULL` elements, the original names will be retained
+      if (!is.null(names) && is.list(dataSets)) {
+        names <- .validateListArgs(names, names(dataSets))
+        names <- ifelse(is.na(names), names(dataSets), names)
+      }
+
+      # they should be of same length
+      if (!is.null(groups)) {
+        validateIsSameLength(dataSets, groups)
+
+        # if a list is specified, convert NULL to NA
+        # and then flatten list to a vector
+        if (is.list(groups)) {
+          groups <- purrr::flatten_chr(purrr::modify(groups, .null_to_na))
+        }
+      }
+
+      # save the original object as it is; useful for `$toDataFrame()` method
+      private$.dataSets <- dataSets
+
+      # save grouping information for observed data
+      private$.groups$groupsDataSets <- groups
+
+      # extract a dataframe and store it internally
+      private$.dataSetsDF <- private$.dataSet2DF(private$.dataSets, groups, names)
+
+      # add it to combined dataframe
+      private$.dataCombinedDF <- private$.updateDF(private$.dataCombinedDF, private$.dataSetsDF)
+
+      if (length(purrr::compact(private$.groups)) > 0L) {
+        private$.groupMap <- private$.extractGroupMap(private$.dataCombinedDF)
+      }
+
+      invisible(self)
+    },
+
+    #' @description
+    #' Transform raw data with required offsets and scale factors.
+    #' @return A dataframe with respective raw quantities plus offsets
+    #'   multiplied by the specified scale factors.
+
+    setDataTransforms = function(names = NULL,
+                                 xOffsets = 0,
+                                 yOffsets = 0,
+                                 xScaleFactors = 1,
+                                 yScaleFactors = 1) {
+      # check that the arguments to parameters make sense
+      validateIsString(names, nullAllowed = TRUE)
+      validateIsNumeric(xOffsets)
+      validateIsNumeric(yOffsets)
+      validateIsNumeric(xScaleFactors)
+      validateIsNumeric(yScaleFactors)
+
+      # styler: off
+      private$.names         <- names
+      private$.xOffsets      <- xOffsets
+      private$.yOffsets      <- yOffsets
+      private$.xScaleFactors <- xScaleFactors
+      private$.yScaleFactors <- yScaleFactors
+      # styler: on
+
+      # transform data
+      if (!is.null(private$.dataCombinedDF)) {
+        private$.dataCombinedDF <- private$.dataTransform(private$.dataCombinedDF)
+      }
+
+      invisible(self)
+    },
+
+    ## getter methods ---------------
+
+    #' @description
+    #' A dataframe of simulated and/or observed data (depending on instances of
+    #' which objects have been added to the object).
+    #'
+    #' Note that the order in which you enter different object matters. If you
+    #' first enter observed data and simulated data later, the rows will also be
+    #' ordered in the same way.
+    #'
+    #' @return A dataframe.
+
+    toDataFrame = function() {
```

Comment:
I am not comfortable with the fact that the method's name is "toDataFrame", but it returns a tibble. I would prefer a data.frame.

Reply:
There are a number of reasons why we should prefer returning a tibble dataframe, as opposed to a classical dataframe, and I would like to get into them during a meeting, but for now, I just want to emphasize that this is not false advertising 😅 

``` r
is.data.frame(tibble::tibble())
#> [1] TRUE
```

<sup>Created on 2022-01-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Reply:
I actually don't know how R can return `TRUE` for this. Tibbles and data.frames have different subsetting syntax, I see it as a huge problem.

Reply:
Yeah, true. But, the good news is that, in most cases, the tibble subsetting behavior is what the users will find intuitive or unsurprising, though.

You can find an extensive list of differences between tibble and dataframe subsetting behaviors here:
https://tibble.tidyverse.org/articles/wip/subset.html